buildengine/deploy.go

```diff
@@ -0,0 +1,230 @@
+package buildengine
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"connectrpc.com/connect"
+	"golang.org/x/exp/maps"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	ftlv1 "github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1"
+	"github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1/ftlv1connect"
+	schemapb "github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1/schema"
+	"github.com/TBD54566975/ftl/common/moduleconfig"
+	"github.com/TBD54566975/ftl/internal/log"
+	"github.com/TBD54566975/ftl/internal/sha256"
+	"github.com/TBD54566975/ftl/internal/slices"
+)
+
+type deploymentArtefact struct {
+	*ftlv1.DeploymentArtefact
+	localPath string
+}
+
+func Deploy(ctx context.Context, module Module, replicas int32, waitForDeployOnline bool, client ftlv1connect.ControllerServiceClient) error {
+	logger := log.FromContext(ctx).Scope(module.Module)
+	ctx = log.ContextWithLogger(ctx, logger)
+	logger.Infof("Deploying module")
+
+	deployDir := filepath.Join(module.Dir, module.DeployDir)
```

Comment:
Unrelated but I keep having to do this, maybe we should add helper methods that do the joining for us, eg. `module.AbsDeployDir()`

Or maybe we should make `DeployDir` et al absolute when we load the config. Not sure.

Reply:
Yeah same! I'll add the `module.AbsDeployDir` for now and maybe we can clean this up with the `DeployDir` stuff if that makes sense